internal/quickfort/dig.lua

```diff
@@ -1,26 +1,569 @@
+any effect.
+]]
 
 if not dfhack_flags.module then
     qerror('this script cannot be called directly')
 end
 
+local utils = require('utils')
 local quickfort_common = reqscript('internal/quickfort/common')
 local log = quickfort_common.log
 
-function do_run(zlevel, grid)
-    local stats = nil
-    print('"quickfort run" not yet implemented for mode: dig')
+local function is_construction(tileattrs)
+    return tileattrs.material == df.tiletype_material.CONSTRUCTION
+end
+
+local function is_floor(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.FLOOR
+end
+
+local function is_diggable_floor(tileattrs)
+    return is_floor(tileattrs) or
+            tileattrs.shape == df.tiletype_shape.BOULDER or
+            tileattrs.shape == df.tiletype_shape.PEBBLES
+end
+
+local function is_wall(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.WALL
+end
+
+local function is_fortification(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.FORTIFICATION
+end
+
+local function is_up_stair(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.STAIR_UP
+end
+
+local function is_down_stair(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.STAIR_DOWN
+end
+
+local function is_removable_shape(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.RAMP or
+            tileattrs.shape == df.tiletype_shape.STAIR_UP or
+            tileattrs.shape == df.tiletype_shape.STAIR_UPDOWN
+end
+
+local function is_gatherable(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.SHRUB
+end
+
+local hard_natural_materials = utils.invert({
+    df.tiletype_material.STONE,
+    df.tiletype_material.FEATURE,
+    df.tiletype_material.LAVA_STONE,
+    df.tiletype_material.MINERAL,
+    df.tiletype_material.FROZEN_LIQUID,
+})
+
+local function is_hard(tileattrs)
+    return hard_natural_materials[tileattrs.material]
+end
+
+local function is_smooth(tileattrs)
+    return tileattrs.special == df.tiletype_special.SMOOTH
+end
+
+local function get_engraving(pos)
+    -- scan through engravings until we find the one at this pos
+    -- super inefficient; is there a better way to do this?
+    for _, engraving in ipairs(df.global.world.engravings) do
+        if same_xyz(pos, engraving.pos) then
+            return engraving
+        end
+    end
+    return nil
+end
+
+local function has_designation(flags, occupancy)
+    return flags.dig ~= df.tile_dig_designation.No or
+            flags.smooth ~= 0 or
+            occupancy.carve_track_north ~= 0 or
+            occupancy.carve_track_east ~= 0 or
+            occupancy.carve_track_south ~= 0 or
+            occupancy.carve_track_west ~= 0
+end
+
+local function clear_designation(flags, occupancy)
+    flags.dig = df.tile_dig_designation.No
+    flags.smooth = 0
+    occupancy.carve_track_north = 0
+    occupancy.carve_track_east = 0
+    occupancy.carve_track_south = 0
+    occupancy.carve_track_west = 0
+end
```

Comment:
I was thinking of suggesting that these be added to the maps module, similar to the existing `getTileDesignation()` and `getTileOccupancy()`. Those both take coordinates, though, although I think your context currently keeps track of that too.

Syncing scripts with core library changes is tricky, of course. Maybe once this is done, we can take a pass through and identify things that would be useful in the core library/modules.

Reply:
agreed. I'll add a comment here to remind us to do that